src/ripple/protocol/impl/TER.cpp

```diff
@@ -42,6 +42,12 @@ transResults()
             TERUnderlyingType,
             std::pair<char const* const, char const* const>> const results
     {
+        MAKE_ERROR(tecAMM_BALANCE,                   "AMM has invalid balance."),
+        MAKE_ERROR(tecAMM_INVALID_TOKENS,            "AMM invalid LP tokens."),
+        MAKE_ERROR(tecAMM_FAILED_DEPOSIT,            "AMM failed deposit."),
+        MAKE_ERROR(tecAMM_FAILED_WITHDRAW,           "AMM failed withdraw."),
+        MAKE_ERROR(tecAMM_FAILED_BID,                "AMM failed bid."),
+        MAKE_ERROR(tecAMM_FAILED_VOTE,               "AMM failed vote."),
```

Comment:
I'm not a huge fan of these four error messages. They don't really tell you anything about what failed, except which transaction failed. And you already know which transaction failed, because that's the transaction that was submitted. In other words, if you're just going to have a generic failure, you may as well replace all of these with `tecFAILED_PROCESSING`.

In most of the places where these codes are being returned, there are log messages explaining what the actual problem is. The problem is that most users are never going to realistically be able to see those log messages, so they're left to guess at what they did wrong.

It might be better to instead create a handful broad categories of failures, and create codes for each of those. e.g.
* Invalid options
* Insufficient balance
* Invalid zero amount
* Insufficient bid
Some of these may be covered by existing codes.

I also noticed that some of the results look like they're returned in conditions that should be impossible, in which case `tecINTERNAL` would be a better choice.

I know I'm being a little vague here, because I am not super familiar with the transactors yet.